apps/nouns-camp/src/hooks/simulation.js

```diff
@@ -0,0 +1,96 @@
+import React from "react";
+import { encodeFunctionData } from "viem";
+
+const SIMULATE_TRANSACTION_TYPES = [
+  "function-call",
+  "proxied-function-call",
+  "payable-function-call",
+  "proxied-payable-function-call",
+  "treasury-noun-transfer",
+  "escrow-noun-transfer",
+  "transfer",
+  "usdc-transfer-via-payer",
+];
+
+export const fetchSimulation = async (transaction) => {
+  if (!SIMULATE_TRANSACTION_TYPES.includes(transaction.type)) {
+    return null;
+  }
+
+  const transactionValue = (transaction) => {
+    switch (transaction.type) {
+      case "payable-function-call":
+      case "proxied-payable-function-call":
+      case "transfer":
+        return transaction.value;
+      default:
+        return 0;
+    }
+  };
+
+  const encodedData = (transaction) => {
+    if (!transaction.functionName) {
+      return "";
+    }
+
+    return encodeFunctionData({
+      abi: [
+        {
+          inputs: transaction.functionInputTypes,
+          name: transaction.functionName,
+          type: "function",
+        },
+      ],
+      functionName: transaction.functionName,
+      args: transaction.functionInputs,
+    });
+  };
+
+  const parsedTransaction = {
+    to: transaction.target,
+    value: transactionValue(transaction).toString(),
+    input: encodedData(transaction),
+  };
+
+  const body = JSON.stringify(parsedTransaction);
+  const res = await fetch("/api/simulate", {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body,
+  });
+
+  if (!res.ok) {
+    throw new Error("Failed to simulate transaction");
+  }
+
+  const simulation = await res.json();
+  return simulation;
+};
+
+export const useTransactionSimulation = (transaction) => {
+  const [isFetching, setIsFetching] = React.useState(false);
+  const [simulation, setSimulation] = React.useState(null);
+
+  const fetchData = React.useCallback(async () => {
+    try {
+      setIsFetching(true);
+      const sim = await fetchSimulation(transaction);
+      setSimulation(sim);
```

Comment:
Might want to add a catch block with setSimulation(null) to prevent showing stale/incorrect data?